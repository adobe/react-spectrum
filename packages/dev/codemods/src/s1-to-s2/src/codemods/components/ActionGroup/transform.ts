import {addComponentImport} from '../../shared/utils';
import {commentOutProp} from '../../shared/transforms';
import {getComponents} from '../../../getComponents';
import {NodePath} from '@babel/traverse';
import * as t from '@babel/types';

let availableComponents = getComponents();

/**
 * Transforms ActionGroup:
 * - Comment out overflowMode (it has not been implemented yet).
 * - Comment out buttonLabelBehavior (it has not been implemented yet).
 * - Comment out summaryIcon (it has not been implemented yet).
 * - Use ActionButtonGroup if no selection.
 * - Use ToggleButtonGroup if selection is used.
 * - Update root level onAction to onPress on each ActionButton.
 * - Apply isDisabled directly on each ActionButton/ToggleButton instead of disabledKeys.
 * - Convert dynamic collections render function to items.map.
 */
export default function transformActionGroup(path: NodePath<t.JSXElement>) {
  // Comment out overflowMode
  commentOutProp(path, {propName: 'overflowMode'});

  // Comment out buttonLabelBehavior
  commentOutProp(path, {propName: 'buttonLabelBehavior'});

  // Comment out summaryIcon
  commentOutProp(path, {propName: 'summaryIcon'});

  let selectionModePath = path.get('openingElement').get('attributes').find((attr) => t.isJSXAttribute(attr.node) && attr.node.name.name === 'selectionMode') as NodePath<t.JSXAttribute> | undefined;
  let selectionMode = t.isStringLiteral(selectionModePath?.node.value) ? selectionModePath.node.value.value : 'none';
  let newComponentName, childComponentName;
  if (selectionMode === 'none') {
    // Use ActionButtonGroup if no selection
    newComponentName = 'ActionButtonGroup';
    childComponentName = 'ActionButton';
    selectionModePath?.remove();
  } else {
    // Use ToggleButtonGroup if selection is used
    newComponentName = 'ToggleButtonGroup';
    childComponentName = 'ToggleButton';
  }

  let localName = newComponentName;
  if (availableComponents.has(newComponentName)) {
    let program = path.findParent((p) => t.isProgram(p.node)) as NodePath<t.Program>;
    localName = addComponentImport(program, newComponentName);
  }

  let localChildName = childComponentName;
  if (availableComponents.has(childComponentName)) {
    let program = path.findParent((p) => t.isProgram(p.node)) as NodePath<t.Program>;
    localChildName = addComponentImport(program, childComponentName);
  }


  // Convert dynamic collection to an array.map.
  let items = path.get('openingElement').get('attributes').find((attr) => t.isJSXAttribute(attr.node) && attr.node.name.name === 'items') as NodePath<t.JSXAttribute> | undefined;
  let itemArg: t.Identifier | undefined;
  if (items && t.isJSXExpressionContainer(items.node.value) && t.isExpression(items.node.value.expression)) {
    let child = path.get('children').find(c => c.isJSXExpressionContainer());
    if (child && child.isJSXExpressionContainer() && t.isFunction(child.node.expression)) {
      let arg = child.node.expression.params[0];
      if (t.isIdentifier(arg)) {
        itemArg = arg;
      }

      child.replaceWith(
        t.jsxExpressionContainer(
          t.callExpression(
            t.memberExpression(
              items.node.value.expression,
              t.identifier('map')
            ),
            [child.node.expression]
          )
        )
      );
    }
  }
  items?.remove();

  let onAction = path.get('openingElement').get('attributes').find((attr) => t.isJSXAttribute(attr.node) && attr.node.name.name === 'onAction') as NodePath<t.JSXAttribute> | undefined;

  // Pull disabledKeys prop out into a variable, converted to a Set.
  // Then we can check it in the isDisabled prop of each item.
  let disabledKeysPath = path.get('openingElement').get('attributes').find((attr) => t.isJSXAttribute(attr.node) && attr.node.name.name === 'disabledKeys') as NodePath<t.JSXAttribute> | undefined;
  let disabledKeys: t.Identifier | undefined;
  if (disabledKeysPath && t.isJSXExpressionContainer(disabledKeysPath.node.value) && t.isExpression(disabledKeysPath.node.value.expression)) {
    disabledKeys = path.scope.generateUidIdentifier('disabledKeys');
    path.scope.push({
      id: disabledKeys,
      init: t.newExpression(t.identifier('Set'), [disabledKeysPath.node.value.expression]),
      kind: 'let'
    });
    disabledKeysPath.remove();
  }

  path.traverse({
    JSXElement(child) {
      if (t.isJSXIdentifier(child.node.openingElement.name) && child.node.openingElement.name.name === 'Item') {
        // Replace Item with ActionButton or ToggleButton.
        let childNode = t.cloneNode(child.node);
        childNode.openingElement.name = t.jsxIdentifier(localChildName);
        if (childNode.closingElement) {
          childNode.closingElement.name = t.jsxIdentifier(localChildName);
        }

        // If there is no key prop and we are using dynamic collections, add a default computed from item.key ?? item.id.
        let key = childNode.openingElement.attributes.find(attr => t.isJSXAttribute(attr) && attr.name.name === 'key') as t.JSXAttribute | undefined;
        if (!key && itemArg) {
          let id = t.jsxExpressionContainer(
            t.logicalExpression(
              '??',
              t.memberExpression(itemArg, t.identifier('key')),
              t.memberExpression(itemArg, t.identifier('id'))
            )
          );

          key = t.jsxAttribute(
            t.jsxIdentifier('key'),
            id
          );

          childNode.openingElement.attributes.push(key);
        }

        // If this is a ToggleButtonGroup, add an id prop in addition to key when needed.
        if (key && newComponentName === 'ToggleButtonGroup') {
          // If we are in an array.map we need both key and id. Otherwise, we only need id.
          if (itemArg) {
            childNode.openingElement.attributes.push(t.jsxAttribute(t.jsxIdentifier('id'), key.value));
          } else {
            key.name.name = 'id';
          }
        }

        let keyValue: t.Expression | undefined = undefined;
        if (key && t.isJSXExpressionContainer(key.value) && t.isExpression(key.value.expression)) {
          keyValue = key.value.expression;
        } else if (key && t.isStringLiteral(key.value)) {
          keyValue = key.value;
        }

        // Add an onPress to each item that calls the previous onAction, passing in the key.
        if (onAction && t.isJSXExpressionContainer(onAction.node.value) && t.isExpression(onAction.node.value.expression)) {
          childNode.openingElement.attributes.push(
            t.jsxAttribute(
              t.jsxIdentifier('onPress'),
              t.jsxExpressionContainer(
                keyValue 
                  ? t.arrowFunctionExpression([], t.callExpression(onAction.node.value.expression, [keyValue]))
                  : onAction.node.value.expression
              )
            )
          );
        }

        // Add an isDisabled prop to each item, testing whether it is in disabledKeys.
        if (disabledKeys && keyValue) {
          childNode.openingElement.attributes.push(
            t.jsxAttribute(
              t.jsxIdentifier('isDisabled'),
              t.jsxExpressionContainer(
                t.callExpression(
                  t.memberExpression(
                    disabledKeys,
                    t.identifier('has')
                  ),
                  [keyValue]
                )
              )
            )
          );
        }

        child.replaceWith(childNode);
      }
    }
  });

  onAction?.remove();

  path.node.openingElement.name = t.jsxIdentifier(localName);
  if (path.node.closingElement) {
    path.node.closingElement.name = t.jsxIdentifier(localName);
  }
} 
