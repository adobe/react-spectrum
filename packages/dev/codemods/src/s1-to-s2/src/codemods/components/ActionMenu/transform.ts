import {commentOutProp} from '../../shared/transforms';
import {NodePath} from '@babel/traverse';
import * as t from '@babel/types';

/**
 * Transforms ActionMenu:
 * - Comment out closeOnSelect (it has not been implemented yet).
 * - Comment out trigger (it has not been implemented yet).
 */
export default function transformActionMenu(path: NodePath<t.JSXElement>) {
  // Comment out closeOnSelect
  commentOutProp(path, {propName: 'closeOnSelect'});

  // Comment out trigger
  commentOutProp(path, {propName: 'trigger'});
} 
