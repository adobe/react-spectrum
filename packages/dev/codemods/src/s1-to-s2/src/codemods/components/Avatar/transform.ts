import {commentOutProp} from '../../shared/transforms';
import {getName} from '../../shared/utils';
import {NodePath} from '@babel/traverse';
import * as t from '@babel/types';

function updateAvatarSize(
  path: NodePath<t.JSXElement>
) {
  if (
    t.isJSXElement(path.node) &&
    t.isJSXIdentifier(path.node.openingElement.name) &&
    getName(path, path.node.openingElement.name) === 'Avatar'
  ) {
    let sizeAttrPath = path.get('openingElement').get('attributes').find((attr) => t.isJSXAttribute(attr.node) && attr.node.name.name === 'size') as NodePath<t.JSXAttribute>;
    if (sizeAttrPath) {
      let value = sizeAttrPath.node.value;
      if (value?.type === 'StringLiteral') {
        const avatarDimensions = {
          'avatar-size-50': 16,
          'avatar-size-75': 18,
          'avatar-size-100': 20,
          'avatar-size-200': 22,
          'avatar-size-300': 26,
          'avatar-size-400': 28,
          'avatar-size-500': 32,
          'avatar-size-600': 36,
          'avatar-size-700': 40
        };
        let val = avatarDimensions[value.value as keyof typeof avatarDimensions];
        if (val != null) {
          sizeAttrPath.node.value = t.jsxExpressionContainer(t.numericLiteral(val));
        }
      }
    }
  }
}

/**
 * Transforms Avatar:
 * - Comment out isDisabled (it has not been implemented yet).
 * - Update size to be a pixel value if it currently matches 'avatar-size-*'.
 */
export default function transformAvatar(path: NodePath<t.JSXElement>) {
  // Comment out isDisabled
  commentOutProp(path, {propName: 'isDisabled'});

  // Update size to be a pixel value if it currently matches 'avatar-size-*'
  updateAvatarSize(path);
} 
