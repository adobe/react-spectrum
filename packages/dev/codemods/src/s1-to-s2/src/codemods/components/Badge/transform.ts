import {NodePath} from '@babel/traverse';
import * as t from '@babel/types';
import {updatePropNameAndValue} from '../../shared/transforms';

/**
 * Transforms Badge:
 * - Change variant="info" to variant="informative".
 */
export default function transformBadge(path: NodePath<t.JSXElement>) {
  // Change variant="info" to variant="informative"
  updatePropNameAndValue(path, {
    oldPropName: 'variant',
    oldPropValue: 'info',
    newPropName: 'variant',
    newPropValue: 'informative'
  });
} 
