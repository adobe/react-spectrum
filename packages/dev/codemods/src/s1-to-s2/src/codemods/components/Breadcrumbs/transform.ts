import {
  addCommentToElement,
  commentOutProp,
  removeProp
} from '../../shared/transforms';
import {NodePath} from '@babel/traverse';
import * as t from '@babel/types';

/**
 * Transforms Breadcrumbs:
 * - Comment out showRoot (it has not been implemented yet).
 * - Comment out isMultiline (it has not been implemented yet).
 * - Comment out autoFocusCurrent (it has not been implemented yet).
 * - Remove size="S" (Small is no longer a supported size in Spectrum 2).
 * - Add comment to wrap in nav element if needed.
 */
export default function transformBreadcrumbs(path: NodePath<t.JSXElement>) {
  // Comment out showRoot
  commentOutProp(path, {propName: 'showRoot'});

  // Comment out isMultiline
  commentOutProp(path, {propName: 'isMultiline'});

  // Comment out autoFocusCurrent
  commentOutProp(path, {propName: 'autoFocusCurrent'});

  // Remove size="S"
  removeProp(path, {propName: 'size', propValue: 'S'});

  // Add comment to wrap in nav element if needed
  addCommentToElement(path, {
    comment: 'S2 Breadcrumbs no longer includes a nav element by default. You can wrap the Breadcrumbs component in a nav element if needed.'
  });
} 
