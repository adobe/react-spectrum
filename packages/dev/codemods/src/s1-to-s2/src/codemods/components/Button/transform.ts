import {NodePath} from '@babel/traverse';
import {
  removeProp,
  updateComponentIfPropPresent,
  updatePropName,
  updatePropNameAndValue,
  updatePropValueAndAddNewPropName
} from '../../shared/transforms';
import * as t from '@babel/types';

/**
 * Transforms Button:
 * - Change variant="cta" to variant="accent"
 * - Change variant="overBackground" to variant="primary" staticColor="white"
 * - Change style to fillStyle
 * - Remove isQuiet (it is no longer supported in Spectrum 2)
 * - If href is present, the Button should be converted to a LinkButton
 * - Remove elementType (it is no longer supported in Spectrum 2).
 */
export default function transformButton(path: NodePath<t.JSXElement>) {
  // Change variant="cta" to variant="accent"
  updatePropNameAndValue(path, {
    oldPropName: 'variant',
    oldPropValue: 'cta',
    newPropName: 'variant',
    newPropValue: 'accent'
  });

  // Change variant="overBackground" to variant="primary" staticColor="white"
  updatePropValueAndAddNewPropName(path, {
    oldPropName: 'variant',
    oldPropValue: 'overBackground',
    newPropName: 'variant',
    newPropValue: 'primary',
    additionalPropName: 'staticColor',
    additionalPropValue: 'white'
  });

  // Change style to fillStyle
  updatePropName(path, {
    oldPropName: 'style',
    newPropName: 'fillStyle'
  });

  // Remove isQuiet
  removeProp(path, {propName: 'isQuiet'});

  // If href is present, the Button should be converted to a LinkButton
  updateComponentIfPropPresent(path, {
    propName: 'href',
    newComponentName: 'LinkButton'
  });

  // Remove elementType
  removeProp(path, {propName: 'elementType'});
}
