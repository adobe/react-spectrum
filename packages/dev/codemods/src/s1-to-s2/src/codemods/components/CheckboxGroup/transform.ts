import {NodePath} from '@babel/traverse';
import {removeProp} from '../../shared/transforms';
import * as t from '@babel/types';

/**
 * Transforms CheckboxGroup:
 * - Remove showErrorIcon (it has been removed due to accessibility issues).
 */
export default function transformCheckboxGroup(path: NodePath<t.JSXElement>) {
  // Remove showErrorIcon
  removeProp(path, {propName: 'showErrorIcon'});
} 
