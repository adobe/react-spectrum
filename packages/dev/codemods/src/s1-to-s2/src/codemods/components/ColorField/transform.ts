import {NodePath} from '@babel/traverse';
import {removeProp, updatePropNameAndValue} from '../../shared/transforms';
import * as t from '@babel/types';

/**
 * Transforms ColorField:
 * - Remove isQuiet (it is no longer supported in Spectrum 2).
 * - Remove placeholder (it has been removed due to accessibility issues).
 * - Change validationState="invalid" to isInvalid.
 * - Remove validationState="valid" (it is no longer supported in Spectrum 2).
 */
export default function transformColorField(path: NodePath<t.JSXElement>) {
  // Remove isQuiet
  removeProp(path, {propName: 'isQuiet'});

  // Remove placeholder
  removeProp(path, {propName: 'placeholder'});

  // Change validationState="invalid" to isInvalid
  updatePropNameAndValue(path, {
    oldPropName: 'validationState',
    oldPropValue: 'invalid',
    newPropName: 'isInvalid',
    newPropValue: true
  });

  // Remove validationState="valid"
  removeProp(path, {propName: 'validationState', propValue: 'valid'});
} 
