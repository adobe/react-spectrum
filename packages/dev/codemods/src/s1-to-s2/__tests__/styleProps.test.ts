/*
 * Copyright 2024 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

// @ts-ignore
import {defineSnapshotTest} from 'jscodeshift/dist/testUtils';
import transform from '../src/codemods/codemod';

const test = (name: string, input: string) => {
  defineSnapshotTest(transform, {}, input, name);
};

test('Handles dimensions', `
import {Button} from '@adobe/react-spectrum';

<Button marginX="size-200" marginY={20} width="100%" minWidth="fit-content">Test</Button>
`);

test('Should keep import aliases', `
import {Button as RSPButton} from '@adobe/react-spectrum';

<RSPButton marginX="size-200" marginY={20} width="100%">Test</RSPButton>
`);

test('Adds a comment when a variable is used', `
import {Button} from '@adobe/react-spectrum';

<Button marginX={margin}>Test</Button>
`);

test('Adds a comment when a spread is used', `
import {Button} from '@adobe/react-spectrum';

<Button marginX="size-200" {...otherProps}>Test</Button>
`);

test('Handles Flex', `
import {Flex} from '@adobe/react-spectrum';

<Flex direction="column" alignItems="center" gap="size-200" wrap>Test</Flex>
`);

test('Handles Grid', `
import {Grid} from '@adobe/react-spectrum';

<Grid
  areas={[
    'header  header',
    'sidebar content',
    'footer  footer'
  ]}
  columns={['1fr', '3fr']}
  rows={['size-1000', 'auto', 'size-1000']}
  height="size-6000"
  gap="size-100">
  Test
</Grid>
`);

test('Handles View', `
import {View} from '@adobe/react-spectrum';

<View
  borderWidth="thin"
  borderColor="dark"
  borderRadius="medium"
  padding="size-250"
  backgroundColor="blue-600"
  flex>
  Test
</View>
`);

test('Handles View colorVersion={5}', `
import {View} from '@adobe/react-spectrum';

<View
  borderWidth="thin"
  borderColor="gray-300"
  borderRadius="medium"
  padding="size-250"
  backgroundColor="blue-600"
  colorVersion={5}>
  Test
</View>
`);

test('Handles View colorVersion={6}', `
import {View} from '@adobe/react-spectrum';

<View
  borderWidth="thin"
  borderColor="gray-300"
  borderRadius="medium"
  padding="size-250"
  backgroundColor="blue-600"
  colorVersion={6}>
  Test
</View>
`);

test('Handles responsive style props', `
import {TextField} from '@adobe/react-spectrum';

<TextField label="Name" width={{ base: 'size-2000', L: 'size-5000' }} />
`);

test('Handles UNSAFE_style', `
import {View} from '@adobe/react-spectrum';

<View
  UNSAFE_style={{
    color: isErrorState ? 'var(--spectrum-red-900)' : 'var(--spectrum-gray-900)',
    borderWidth: border === 'thin' ? 'var(--spectrum-alias-border-size-thin)' : 'var(--spectrum-alias-border-size-thick)',
    borderRadius: isRounded(props) ? 4 : 0,
    display: 'flex',
    gap: 'var(--spectrum-global-dimension-size-200)',
    fontSize: 'var(--spectrum-global-dimension-font-size-100)',
    fontWeight: 'var(--spectrum-global-font-weight-bold)',
    gridTemplateAreas: '"foo bar" "baz qux"',
    transitionDuration: 'var(--spectrum-global-animation-duration-200)'
  }}>
  Test
</View>
`);

test('Bails on UNSAFE_style if it contains a spread', `
import {View} from '@adobe/react-spectrum';

<View
  UNSAFE_style={{
    color: 'var(--spectrum-gray-900)',
    ...otherStyles
  }}>
  Test
</View>
`);

test('Adds a comment to UNSAFE_className', `
import {Button} from '@adobe/react-spectrum';

<Button UNSAFE_className="hi">
  Test
</Button>
`);

test('Converts UNSAFE_className to className on View', `
import {View} from '@adobe/react-spectrum';

<View UNSAFE_className="hi">
  Test
</View>
`);

test('Merges UNSAFE_className with macro className', `
import {View} from '@adobe/react-spectrum';

<View UNSAFE_className="hi" marginX="size-200">
  Test
</View>
`);

test('Handles responsive style props', `
import {TextField} from '@adobe/react-spectrum';

<TextField label="Name" width={{ base: 'size-2000', L: 'size-5000' }} />
`);

test('Leaves comment for invalid style props', `
import {TextField} from '@adobe/react-spectrum';

<TextField label="Name" width="250" />
`);

test('Updates left/right layout values to be start/end', `
import {Button, Flex} from '@adobe/react-spectrum';

<>
  <Flex alignContent="left">Test</Flex>
  <Flex alignItems="left">Test</Flex>
  <Flex justifyContent="left">Test</Flex>
  <Flex justifyItems="left">Test</Flex>
  <Button alignSelf="left">Test</Button>
  <Button justifySelf="left">Test</Button>

  <Flex alignContent="right">Test</Flex>
  <Flex alignItems="right">Test</Flex>
  <Flex justifyContent="right">Test</Flex>
  <Flex justifyItems="right">Test</Flex>
  <Button alignSelf="right">Test</Button>
  <Button justifySelf="right">Test</Button>
</>
`);

test('Should leave existing style macros unaffected', `  
import { TextField } from "@react-spectrum/s2";
import { style } from "@react-spectrum/s2/style" with { type: "macro" };

<TextField label="Name" styles={style({width: 160})} />
`);

test('Handles multi-value margin/padding shorthands', `
import {View} from '@adobe/react-spectrum';

<>
  <View margin="10px">Single Value Margin</View>

  <View margin="10px 16px">Two Value Margin</View>

  <View margin="10px 16px 20px">Three Value Margin</View>

  <View margin="10px 16px 20px 24px">Four Value Margin</View>

  <View padding="10px">Single Value Padding</View>

  <View padding="10px 16px">Two Value Padding</View>

  <View padding="10px 16px 20px">Three Value Padding</View>

  <View padding="10px 16px 20px 24px">Four Value Padding</View>
</>
`);
