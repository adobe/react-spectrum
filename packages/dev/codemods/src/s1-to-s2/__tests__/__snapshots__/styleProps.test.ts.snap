// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adds a comment to UNSAFE_className 1`] = `
"import { Button } from "@react-spectrum/s2";

<Button // TODO(S2-upgrade): check this UNSAFE_className
UNSAFE_className="hi">
  Test
</Button>"
`;

exports[`Adds a comment when a spread is used 1`] = `
"import { Button } from "@react-spectrum/s2";
import { style } from "@react-spectrum/s2/style" with { type: "macro" };

<Button
  // TODO(S2-upgrade): check this spread for style props
  {...otherProps}
  styles={style({
    marginX: 16
  })}>Test</Button>"
`;

exports[`Adds a comment when a variable is used 1`] = `
"import { Button } from "@react-spectrum/s2";

<Button // TODO(S2-upgrade): update this style prop
marginX={margin}>Test</Button>"
`;

exports[`Bails on UNSAFE_style if it contains a spread 1`] = `
"<div
  // TODO(S2-upgrade): check this UNSAFE_style
  UNSAFE_style={{
    color: 'var(--spectrum-gray-900)',
    ...otherStyles
  }}>
  Test
</div>"
`;

exports[`Converts UNSAFE_className to className on View 1`] = `
"<div className="hi">
  Test
</div>"
`;

exports[`Handles Flex 1`] = `
"import { style } from "@react-spectrum/s2/style" with { type: "macro" };

<div
  className={style({
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    gap: 16,
    flexWrap: "wrap"
  })}>Test</div>"
`;

exports[`Handles Grid 1`] = `
"import { style } from "@react-spectrum/s2/style" with { type: "macro" };

<div
  className={style({
    display: "grid",
    gridTemplateAreas: ["header  header", "sidebar content", "footer  footer"],
    gridTemplateColumns: ["1fr", "3fr"],
    gridTemplateRows: [80, "auto", 80],
    height: 480,
    gap: 8
  })}>
  Test
</div>"
`;

exports[`Handles UNSAFE_style 1`] = `
"import { style } from "@react-spectrum/s2/style" with { type: "macro" };

<div
  className={style({
    color: {
      default: "gray-900",
      isErrorState: "red-900"
    },

    borderWidth: {
      default: 2,

      border: {
        thin: 1
      }
    },

    borderRadius: {
      default: "none",
      isRounded: "default"
    },

    display: "flex",
    gap: 16,
    fontSize: "ui",
    fontWeight: "bold",
    gridTemplateAreas: ["foo bar", "baz qux"],
    transitionDuration: 160
  })({
    isErrorState,
    border,
    isRounded: isRounded(props)
  })}>
  Test
</div>"
`;

exports[`Handles View 1`] = `
"import { style, lightDark } from "@react-spectrum/s2/style" with { type: "macro" };

<div
  className={style({
    borderWidth: 1,
    borderColor: "gray-400",
    borderRadius: "lg",
    padding: 20,
    backgroundColor: lightDark("blue-1000", "blue-900"),
    flex: 1
  })}>
  Test
</div>"
`;

exports[`Handles View colorVersion={5} 1`] = `
"import { style, lightDark } from "@react-spectrum/s2/style" with { type: "macro" };

<div
  className={style({
    borderWidth: 1,
    borderColor: "gray-200",
    borderRadius: "lg",
    padding: 20,
    backgroundColor: lightDark("blue-1000", "blue-900")
  })}>
  Test
</div>"
`;

exports[`Handles View colorVersion={6} 1`] = `
"import { style } from "@react-spectrum/s2/style" with { type: "macro" };

<div
  className={style({
    borderWidth: 1,
    borderColor: "gray-200",
    borderRadius: "lg",
    padding: 20,
    backgroundColor: "blue-600"
  })}>
  Test
</div>"
`;

exports[`Handles dimensions 1`] = `
"import { Button } from "@react-spectrum/s2";
import { style } from "@react-spectrum/s2/style" with { type: "macro" };

<Button
  styles={style({
    marginX: 16,
    marginY: 20,
    width: "full",
    minWidth: "fit"
  })}>Test</Button>"
`;

exports[`Handles multi-value margin/padding shorthands 1`] = `
"import { style } from "@react-spectrum/s2/style" with { type: "macro" };

<>
  <div className={style({
    margin: "[10px]"
  })}>Single Value Margin</div>

  <div className={style({
    marginY: "[10px]",
    marginX: 16
  })}>Two Value Margin</div>

  <div className={style({
    marginTop: "[10px]",
    marginX: 16,
    marginBottom: 20
  })}>Three Value Margin</div>

  <div className={style({
    marginTop: "[10px]",
    marginRight: 16,
    marginBottom: 20,
    marginLeft: 24
  })}>Four Value Margin</div>

  <div className={style({
    padding: "[10px]"
  })}>Single Value Padding</div>

  <div className={style({
    paddingY: "[10px]",
    paddingX: 16
  })}>Two Value Padding</div>

  <div className={style({
    paddingTop: "[10px]",
    paddingX: 16,
    paddingBottom: 20
  })}>Three Value Padding</div>

  <div className={style({
    paddingTop: "[10px]",
    paddingRight: 16,
    paddingBottom: 20,
    paddingLeft: 24
  })}>Four Value Padding</div>
</>"
`;

exports[`Handles responsive style props 1`] = `
"import { TextField } from "@react-spectrum/s2";
import { style } from "@react-spectrum/s2/style" with { type: "macro" };

<TextField label="Name" styles={style({
  width: {
    default: 160,
    lg: 400
  }
})} />"
`;

exports[`Handles responsive style props 2`] = `
"import { TextField } from "@react-spectrum/s2";
import { style } from "@react-spectrum/s2/style" with { type: "macro" };

<TextField label="Name" styles={style({
  width: {
    default: 160,
    lg: 400
  }
})} />"
`;

exports[`Leaves comment for invalid style props 1`] = `
"import { TextField } from "@react-spectrum/s2";

// TODO(S2-upgrade): Could not transform style prop automatically: Error: invalid dimension: 250
<TextField label="Name" width="250" />"
`;

exports[`Merges UNSAFE_className with macro className 1`] = `
"import { style } from "@react-spectrum/s2/style" with { type: "macro" };

<div
  className={"hi" + style({
    marginX: 16
  })}>
  Test
</div>"
`;

exports[`Should keep import aliases 1`] = `
"import { Button as RSPButton } from "@react-spectrum/s2";
import { style } from "@react-spectrum/s2/style" with { type: "macro" };

<RSPButton
  styles={style({
    marginX: 16,
    marginY: 20,
    width: "full"
  })}>Test</RSPButton>"
`;

exports[`Should leave existing style macros unaffected 1`] = `
"import { TextField } from "@react-spectrum/s2";
import { style } from "@react-spectrum/s2/style" with { type: "macro" };

<TextField label="Name" styles={style({width: 160})} />"
`;

exports[`Updates left/right layout values to be start/end 1`] = `
"import { Button } from "@react-spectrum/s2";
import { style } from "@react-spectrum/s2/style" with { type: "macro" };

<>
  <div className={style({
    display: "flex",
    alignContent: "start"
  })}>Test</div>
  <div className={style({
    display: "flex",
    alignItems: "start"
  })}>Test</div>
  <div className={style({
    display: "flex",
    justifyContent: "start"
  })}>Test</div>
  <div
    justifyItems="start"
    className={style({
      display: "flex"
    })}>Test</div>
  <Button styles={style({
    alignSelf: "start"
  })}>Test</Button>
  <Button styles={style({
    justifySelf: "start"
  })}>Test</Button>

  <div className={style({
    display: "flex",
    alignContent: "end"
  })}>Test</div>
  <div className={style({
    display: "flex",
    alignItems: "end"
  })}>Test</div>
  <div className={style({
    display: "flex",
    justifyContent: "end"
  })}>Test</div>
  <div
    justifyItems="end"
    className={style({
      display: "flex"
    })}>Test</div>
  <Button styles={style({
    alignSelf: "end"
  })}>Test</Button>
  <Button styles={style({
    justifySelf: "end"
  })}>Test</Button>
</>"
`;
